Submit shipments sql fix

Levia opened the pull request:
This is related to [Submission into Trade DB](https://www.pivotaltracker.com/story/show/132315085).
Due to the amendments made to the ARU model, this sql scripts needs to be updated in order to handle nullable fields such as `updated_at`, `epix_updated_at`, `updated_by_id`, `epix_updated_by_id`.
The `updated_by` field gets populated based on the fact that is not null and also depending on which of the `updated_at` fields is the latest.

agnessa commented:
I don't think this will do for updated_by_id and created_by_id, because those fields have foreign keys constraints referencing the users table, so we can't just push epix ids inside, there's no way to interpret that data. The thing to do would be to remove the not null constraint on those and consider how to accommodate epix ids separately.

Levia commented:
I've updated the PR adding what it needs to handle epix timestamps and created/updated by.
I've also added then the required fields in the trade_sandbox_template table and removing the not null constraint to the default timestamps.
Unfortunately I'm having issues understanding why the Travis build fails when creating the sql function.
This PR is required for the trade_reporting_tool build to pass for the `submission_into_trade_db` branch.

agnessa commented:
When you change the signature of an SQL function, you must drop the old function and then create a new one. The reason it is not failing for you locally is because you have 2 versions of the function defined, whereas travis only has the last version and hits a mismatch between that version and the `COMMENT` clause which doesn't match.

agnessa commented:
You only want one version of the function in the db, so a `DROP FUNCTION` statement is necessary in the script / migrationto remove the old version from staging / production. Also, once that is changed, the Rails code related to submission must be amended so that submission keeps working as previously via the Species+ tool.

Levia commented:
Thank you for your hints! Hopefully it's looking better now :D

agnessa commented on db/plpgsql/011_copy_transactions_from_sandbox_to_shipments.sql:
```diff
@@ -165,29 +191,45 @@ BEGIN
-        current_timestamp, ' ||
-        aru.created_by_id || ', ' ||
-        aru.updated_by_id || '
+        NULLIF(''' || created_at || '''::timestamp, ''' || to_timestamp(0) || '''::timestamp),
```
I read this check roughly as: `if created_at == 1970-01-01 then null else created_at`; but why would `created_at` have that value, does it really happen? :scream_cat:

agnessa replied:
Actually maybe this is not relevant, I keep forgetting: these timestamps should be NOW(), not whatever was the aru creation date. The thing to check is which timestamp to set using the `sapi_type` I guess?

Levia replied:
It might be that I have considered that wrongly. My thought was to keep the timestamps as they are in the sandbox template, so if those are already set, then I need to copy those as well, that's why I keep checking if those are `NULL` use `NULL` otherwise keep that one that's in there already. This complicates the whole thing. 
If, instead, we want to create the records in the shipments table, without taking into consideration the timestamps and the `created_by` and `updated_by` in the sandbox table, than I guess it's much easier that this. If that's the case, it should be as easy as you say for both timestamps and ids

Levia replied:
Cool, than I guess we can do the same of the ids

agnessa commented on db/plpgsql/011_copy_transactions_from_sandbox_to_shipments.sql:
```diff
@@ -165,29 +191,45 @@ BEGIN
+        NULLIF(''' || epix_created_at || '''::timestamp, ''' || to_timestamp(0) || '''::timestamp),
+        NULLIF(''' || epix_updated_at || '''::timestamp, ''' || to_timestamp(0) || '''::timestamp),
+        CASE WHEN ' || sapi_type || ' IS TRUE THEN ' || submitter_id || '
+             ELSE NULLIF(' || COALESCE(aru.created_by_id, 0) ||', 0)
```
isn't this a bit too complicated? if this is a sapi submitter, store the submitter_id, otherwise just store null?

Levia replied:
That's definitely a yes if we don't need to copy over information from the sandbox_table. What I mean is that if we don't care about all those creation and updates that happened in the sandbox, than we can do as you say, otherwise that's why I did all that 😅

agnessa commented on db/plpgsql/011_copy_transactions_from_sandbox_to_shipments.sql:
```diff
@@ -144,6 +152,24 @@ BEGIN
   RAISE INFO '[%] Inserted % permits', table_name, inserted_rows;
 
+  sapi_type := CASE WHEN submitter_type = 'Sapi' THEN true ELSE false END;
+  created_at := COALESCE(aru.created_at, to_timestamp(0));
```
ok so I'm assuming this is because we have a not null constraint on `created_at` - but since we're changing the time-stamping, perhaps it's easier to relax the constraint than to force the data (1970) into it like that?

agnessa commented:
small fixes to stuff I spotted when trying to submit an epix-generated aru via Species+

Levia commented:
Can I push my suggested changes to the SQL function?

Levia commented:
I've pushed the changes to the SQL function as well as a fix for the specs.
So, with the new changes, we will lose information on about who and when some changes have been made on the same shipment in the sandbox and we just store who did the submission and when the it happened.
